cache/cache.go

```diff
@@ -0,0 +1,144 @@
+package cache
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+)
+
+const NodeSize = 32
+
+type Writer struct {
+	*cache
+}
+
+func NewWriterWithLayerFactories(layerFactories []LayerFactory) *Writer {
+	return &Writer{
+		cache: &cache{
+			layers:         make(map[uint]LayerReadWriter),
+			layerFactories: layerFactories,
+		},
+	}
+}
+
+func (c *Writer) GetLayerWriter(layer uint) LayerWriter {
+	layerReadWriter, found := c.layers[layer]
+	if !found {
+		layerReadWriter = c.fillLayerIfNeeded(layer)
+	}
+	return layerReadWriter
+}
+
+func (c *Writer) SetHash(hashFunc func(lChild, rChild []byte) []byte) {
+	c.hash = hashFunc
+}
+
+func (c *Writer) GetReader() (*Reader, error) {
+	err := c.ValidateStructure()
```

Comment:
Not sure I understand the question. If you look at the flow I described above, `GetReader` should only be called once, after the tree is constructed to obtain the reader. Since the reader is read-only, there's no need to validate the structure more than once, upon reader creation.

Reply:
I misunderstood the way it will be used. I thought that it might be called more than once and in such case, you should cache the validation result